runtime/lua/vim/cache.lua

```diff
@@ -0,0 +1,481 @@
+    }
+  end
+  M.track('read', start)
+end
+
+--- The `package.loaders` loader for lua files using the cache.
+---@param modname string module name
+---@return string|function
+---@private
+function Cache.loader(modname)
+  local start = uv.hrtime()
+  local modpath, hash = M.find(modname)
+  if modpath then
+    local chunk, err = M.load(modpath, { hash = hash })
+    M.track('loader', start)
+    return chunk or error(err)
+  end
+  M.track('loader', start)
+  return '\ncache_loader: module ' .. modname .. ' not found'
+end
+
+--- The `package.loaders` loader for libs
+---@param modname string module name
+---@return string|function
+---@private
+function Cache.loader_lib(modname)
+  local start = uv.hrtime()
+  local modpath = M.find(modname, { patterns = jit.os:find('Windows') and { '.dll' } or { '.so' } })
+  ---@type function?, string?
+  if modpath then
+    -- Making function name in Lua 5.1 (see src/loadlib.c:mkfuncname) is
+    -- a) strip prefix up to and including the first dash, if any
+    -- b) replace all dots by underscores
+    -- c) prepend "luaopen_"
+    -- So "foo-bar.baz" should result in "luaopen_bar_baz"
+    local dash = modname:find('-', 1, true)
+    local funcname = dash and modname:sub(dash + 1) or modname
+    local chunk, err = package.loadlib(modpath, 'luaopen_' .. funcname:gsub('%.', '_'))
+    M.track('loader_lib', start)
+    return chunk or error(err)
+  end
+  M.track('loader_lib', start)
+  return '\ncache_loader_lib: module ' .. modname .. ' not found'
+end
+
+--- `loadfile` using the cache
+---@param filename? string
+---@param mode? "b"|"t"|"bt"
+---@param env? table
+---@return function?, string?  error_message
+---@private
+-- luacheck: ignore 312
+function Cache.loadfile(filename, mode, env)
+  local start = uv.hrtime()
+  filename = Cache.normalize(filename)
+  mode = nil -- ignore mode, since we byte-compile the lua source files
+  local chunk, err = M.load(filename, { mode = mode, env = env })
+  M.track('loadfile', start)
+  return chunk, err
+end
+
+--- Checks whether two cache hashes are the same based on:
+--- * file size
+--- * mtime in seconds
+--- * mtime in nanoseconds
+---@param h1 CacheHash
+---@param h2 CacheHash
+---@private
+function Cache.eq(h1, h2)
+  return h1
+    and h2
+    and h1.size == h2.size
+    and h1.mtime.sec == h2.mtime.sec
+    and h1.mtime.nsec == h2.mtime.nsec
+end
+
+--- Loads the given module path using the cache
+---@param modpath string
+---@param opts? {hash?: CacheHash, mode?: "b"|"t"|"bt", env?:table} (table|nil) Options for loading the module:
+---    - hash: (table) the hash of the file to load if it is already known. (defaults to `vim.loop.fs_stat({modpath})`)
+---    - mode: (string) the mode to load the module with. "b"|"t"|"bt" (defaults to `nil`)
+---    - env: (table) the environment to load the module in. (defaults to `nil`)
+---@see |luaL_loadfile()|
+---@return function?, string? error_message
+function M.load(modpath, opts)
```

Comment:
Private?

Reply:
used by lazy, so I don't think it's a good idea to make this private